internal/dkg/actions_passive.go

```diff
@@ -2,130 +2,133 @@ package dkg
 
 import (
 	"context"
+	"encoding/hex"
 	"errors"
-	"time"
+	"fmt"
 
 	"github.com/drand/drand/internal/metrics"
+	"github.com/drand/drand/internal/util"
 	"github.com/drand/drand/protobuf/drand"
 )
 
 // actions_passive contains all internal messaging between nodes triggered by the protocol - things it does automatically
 // upon receiving messages from other nodes: storing proposals, aborting when the leader aborts, etc
 
-func (d *Process) Propose(ctx context.Context, proposal *drand.ProposalTerms) (*drand.EmptyResponse, error) {
-	_, span := metrics.NewSpan(ctx, "dkg.Propose")
-	defer span.End()
+func (d *Process) Packet(ctx context.Context, packet *drand.GossipPacket) (*drand.EmptyDKGResponse, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
 
```

Comment:
do we really need the lock? And should it be a write lock or a read lock?

Reply:
we need a write lock for a few reasons:
- we cache a hash of packets so we don't process the same one multiple times; without a lock we might try and process a packet twice
- we want to process packets one-by-one. Ideally we wouldn't e.g. receive an acceptance before receiving a proposal. This doesn't _strictly_ stop that happening, but it does stop us fetching DB state that becomes stale during processing.

In theory, we could do finer locking than this (on specific areas or by `beaconID`), though it's not a hotspot and there are always dragons when trying to do finer grained locking. If we identify it's a performance issue we can look deeper.